dll/win32/shell32/shlfolder.cpp

```diff
@@ -266,59 +266,80 @@ HRESULT SHELL32_CompareDetails(IShellFolder2* isf, LPARAM lParam, LPCITEMIDLIST
     return MAKE_COMPARE_HRESULT(ret);
 }
 
-void AddClassKeyToArray(const WCHAR * szClass, HKEY* array, UINT* cKeys)
+LSTATUS AddClassKeyToArray(const WCHAR * szClass, HKEY* array, UINT* cKeys)
 {
     if (*cKeys >= 16)
-        return;
+        return ERROR_MORE_DATA;
 
     HKEY hkey;
     LSTATUS result = RegOpenKeyExW(HKEY_CLASSES_ROOT, szClass, 0, KEY_READ | KEY_QUERY_VALUE, &hkey);
-    if (result != ERROR_SUCCESS)
-        return;
-
-    array[*cKeys] = hkey;
-    *cKeys += 1;
+    if (result == ERROR_SUCCESS)
+    {
+        array[*cKeys] = hkey;
+        *cKeys += 1;
+    }
+    return result;
 }
 
-void AddFSClassKeysToArray(PCUITEMID_CHILD pidl, HKEY* array, UINT* cKeys)
+void AddFSClassKeysToArray(UINT cidl, PCUITEMID_CHILD_ARRAY apidl, HKEY* array, UINT* cKeys)
 {
+    //learn.microsoft.com/en-us/windows/win32/shell/fa-associationarray#about-association-arrays
+
+    ASSERT(cidl >= 1 && apidl);
+    PCUITEMID_CHILD pidl = apidl[0];
     if (_ILIsValue(pidl))
     {
+        WCHAR buf[MAX_PATH], *name;
         FileStructW* pFileData = _ILGetFileStructW(pidl);
```

Comment:
I know this is existing code, but this means that this function can _only_ be called from a shell folder that uses `FileStructW` for a pidl.
The wine shell code uses this in all kinds of places, and that is plain wrong.

Reply:
> this means that this function can _only_ be called from a shell folder that uses `FileStructW` for a pidl

Mostly correct.

This function being named `AddFSClassKeysToArray` should make it clear that it can only be used by callers that "are" SFGAO_FILESYSTEM and have compatible pidls.

The old code only supported FileStructW pidls but I added a fallback to _ILSimpleGetTextW so it will at least be compatible with most file pidls now.

> The wine shell code uses this in all kinds of places, and that is plain wrong.

Yes but I don't know how much choice we have. It is internal to shell32 at least.

Non-FS folders are supposed to hand out a custom `IQueryAssociations`.